Error while retrieving specific Deploy

Hi ! 👋

It seems that some deploys, at least the `2f430047c90597ef30467bb5393081e83a69c76ae215b1a89e01918748831a60` and `f8c3d8a9bc4b2a07928c9eeaa8bac41ec54b13cc767eb25509770b333f0f77e9` (WASM deploy), are not well handled by the JSON deserializer.
Tested with the release `2.3.3`

```java
DeployData deploy = casperService.getDeploy("2f430047c90597ef30467bb5393081e83a69c76ae215b1a89e01918748831a60");
```
returns the following exception:

```
java.lang.reflect.UndeclaredThrowableException
	at ProcessorTest.testDeployProcessor(ProcessorTest.java:20)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Index 0 out of bounds for length 0
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: com.casper.sdk.model.deploy.DeployData["execution_results"]->java.util.ArrayList[0]->com.casper.sdk.model.deploy.JsonExecutionResult["result"]->com.casper.sdk.model.deploy.executionresult.Success["effect"]->com.casper.sdk.model.deploy.ExecutionEffect["transforms"]->java.util.ArrayList[11]->com.casper.sdk.model.deploy.Entry["transform"]->com.casper.sdk.model.deploy.transform.WriteCLValue["WriteCLValue"]->com.casper.sdk.model.clvalue.CLValueList["bytes"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:276)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:630)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:618)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:143)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.casper.sdk.jackson.deserializer.AbstractAnyOfDeserializer.deserializeTypedFromObject(AbstractAnyOfDeserializer.java:61)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.casper.sdk.jackson.deserializer.AbstractAnyOfDeserializer.deserializeTypedFromObject(AbstractAnyOfDeserializer.java:61)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer._deserialize(AsWrapperTypeDeserializer.java:121)
	at com.fasterxml.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer.deserializeTypedFromObject(AsWrapperTypeDeserializer.java:52)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4801)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3043)
	at com.googlecode.jsonrpc4j.JsonRpcClient.constructResponseObject(JsonRpcClient.java:292)
	at com.googlecode.jsonrpc4j.JsonRpcClient.readResponse(JsonRpcClient.java:202)
	at com.googlecode.jsonrpc4j.JsonRpcClient.readResponse(JsonRpcClient.java:539)
	at com.googlecode.jsonrpc4j.JsonRpcHttpClient.invoke(JsonRpcHttpClient.java:143)
	at com.googlecode.jsonrpc4j.ProxyUtil$3.invoke(ProxyUtil.java:213)
	... 1 more
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at com.casper.sdk.model.clvalue.CLValueList.setChildTypes(CLValueList.java:103)
	at com.casper.sdk.model.clvalue.CLValueList.serialize(CLValueList.java:59)
	at com.casper.sdk.model.clvalue.serde.CasperSerializableObject.serialize(CasperSerializableObject.java:36)
	at com.casper.sdk.model.clvalue.AbstractCLValue.setValue(AbstractCLValue.java:54)
	at com.casper.sdk.model.clvalue.CLValueList.deserializeCustom(CLValueList.java:98)
	at com.casper.sdk.model.clvalue.AbstractCLValue.deserialize(AbstractCLValue.java:124)
	at com.casper.sdk.model.clvalue.AbstractCLValueWithChildren.setJsonBytes(AbstractCLValueWithChildren.java:53)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:141)
	... 42 more


Process finished with exit code -1
```

```java
DeployData deploy = casperService.getDeploy("f8c3d8a9bc4b2a07928c9eeaa8bac41ec54b13cc767eb25509770b333f0f77e9");
```
returns the following exception:
```
java.lang.reflect.UndeclaredThrowableException
	at ProcessorTest.testDeployProcessor(ProcessorTest.java:15)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: N/A
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: com.casper.sdk.model.deploy.DeployData["deploy"]->com.casper.sdk.model.deploy.Deploy["session"]->com.casper.sdk.model.deploy.executabledeploy.ModuleBytes["args"]->java.util.ArrayList[2]->com.casper.sdk.model.deploy.NamedArg["clValue"]->com.casper.sdk.model.clvalue.CLValueMap["bytes"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:276)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:630)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:618)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:143)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.casper.sdk.jackson.deserializer.AbstractAnyOfDeserializer.deserializeTypedFromObject(AbstractAnyOfDeserializer.java:61)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
	at com.fasterxml.jackson.databind.deser.impl.BeanAsArrayDeserializer.deserialize(BeanAsArrayDeserializer.java:127)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer._deserialize(AsWrapperTypeDeserializer.java:121)
	at com.fasterxml.jackson.databind.jsontype.impl.AsWrapperTypeDeserializer.deserializeTypedFromObject(AsWrapperTypeDeserializer.java:52)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4801)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3043)
	at com.googlecode.jsonrpc4j.JsonRpcClient.constructResponseObject(JsonRpcClient.java:292)
	at com.googlecode.jsonrpc4j.JsonRpcClient.readResponse(JsonRpcClient.java:202)
	at com.googlecode.jsonrpc4j.JsonRpcClient.readResponse(JsonRpcClient.java:539)
	at com.googlecode.jsonrpc4j.JsonRpcHttpClient.invoke(JsonRpcHttpClient.java:143)
	at com.googlecode.jsonrpc4j.ProxyUtil$3.invoke(ProxyUtil.java:213)
	... 1 more
Caused by: java.util.NoSuchElementException
	at com.casper.sdk.model.clvalue.CLValueMap.setChildTypes(CLValueMap.java:116)
	at com.casper.sdk.model.clvalue.CLValueMap.serialize(CLValueMap.java:60)
	at com.casper.sdk.model.clvalue.serde.CasperSerializableObject.serialize(CasperSerializableObject.java:36)
	at com.casper.sdk.model.clvalue.AbstractCLValue.setValue(AbstractCLValue.java:54)
	at com.casper.sdk.model.clvalue.CLValueMap.deserializeCustom(CLValueMap.java:110)
	at com.casper.sdk.model.clvalue.AbstractCLValue.deserialize(AbstractCLValue.java:124)
	at com.casper.sdk.model.clvalue.AbstractCLValueWithChildren.setJsonBytes(AbstractCLValueWithChildren.java:53)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:141)
	... 30 more


Process finished with exit code -1
```

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueList.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueList.java
@@ -108,7 +108,11 @@ public class CLValueList extends AbstractCLValueWithChildren<List<? extends Abst
 
     @Override
     protected void setChildTypes(List<? extends AbstractCLValue<?, ?>> value) {
-        clType.setListType(value.get(0).getClType());
+        if (!value.isEmpty()) {
+            clType.setListType(value.get(0).getClType());
+        } else {
+            clType.setChildTypes(new ArrayList<>());
+        }
     }
 
     @Override

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueMap.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueMap.java
@@ -16,6 +16,7 @@ import lombok.NoArgsConstructor;
 import lombok.Setter;
 import org.bouncycastle.util.encoders.Hex;
 
+import java.util.ArrayList;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -124,8 +125,12 @@ public class CLValueMap extends
     @Override
     @JsonIgnore
     protected void setChildTypes(Map<? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>> value) {
-        Entry<? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>> entry = value.entrySet().iterator().next();
-        clType.setKeyValueTypes(new CLTypeMap.CLTypeMapEntryType(entry.getKey().getClType(), entry.getValue().getClType()));
+        if (value != null && value.entrySet().iterator().hasNext()) {
+            Entry<? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>> entry = value.entrySet().iterator().next();
+            clType.setKeyValueTypes(new CLTypeMap.CLTypeMapEntryType(entry.getKey().getClType(), entry.getValue().getClType()));
+        } else {
+            clType.setChildTypes(new ArrayList<>());
+        }
     }
 
     // This needed to be customized to ensure equality is being checked correctly.

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueOption.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueOption.java
@@ -105,7 +105,11 @@ public class CLValueOption extends AbstractCLValueWithChildren<Optional<Abstract
 
     @Override
     protected void setChildTypes(Optional<AbstractCLValue<?, ?>> value) {
-        clType.setOptionType(value.isPresent() ? value.get().getClType() : null);
+        if (value.isPresent()) {
+            clType.setOptionType(value.get().getClType());
+        } else {
+            clType.setChildTypes(null);
+        }
     }
 
     @Override

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple1.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple1.java
@@ -17,6 +17,7 @@ import lombok.Setter;
 import org.bouncycastle.util.encoders.Hex;
 import org.javatuples.Unit;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 
 /**
@@ -88,7 +89,11 @@ public class CLValueTuple1 extends AbstractCLValueWithChildren<Unit<? extends Ab
 
     @Override
     protected void setChildTypes(Unit<? extends AbstractCLValue<?, ?>> value) {
-        clType.setChildTypes(Arrays.asList(value.getValue0().getClType()));
+        if (value.getValue0() != null) {
+            clType.setChildTypes(Arrays.asList(value.getValue0().getClType()));
+        } else {
+            clType.setChildTypes(new ArrayList<>());
+        }
     }
 
     @Override

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple2.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple2.java
@@ -17,6 +17,7 @@ import lombok.Setter;
 import org.bouncycastle.util.encoders.Hex;
 import org.javatuples.Pair;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 
 /**
@@ -100,7 +101,11 @@ public class CLValueTuple2
 
     @Override
     protected void setChildTypes(Pair<? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>> value) {
-        clType.setChildTypes(Arrays.asList(value.getValue0().getClType(), value.getValue1().getClType()));
+        if (value.getValue0() != null && value.getValue1() != null) {
+            clType.setChildTypes(Arrays.asList(value.getValue0().getClType(), value.getValue1().getClType()));
+        } else {
+            clType.setChildTypes(new ArrayList<>());
+        }
     }
 
     @Override

--- a/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple3.java
+++ b/src/main/java/com/casper/sdk/model/clvalue/CLValueTuple3.java
@@ -17,6 +17,7 @@ import lombok.Setter;
 import org.bouncycastle.util.encoders.Hex;
 import org.javatuples.Triplet;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 
 /**
@@ -112,8 +113,12 @@ public class CLValueTuple3 extends
 
     @Override
     protected void setChildTypes(Triplet<? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>, ? extends AbstractCLValue<?, ?>> value) {
-        clType.setChildTypes(Arrays.asList(value.getValue0().getClType(), value.getValue1().getClType(),
-                value.getValue2().getClType()));
+        if (value.getValue0() != null && value.getValue1() != null && value.getValue2() != null) {
+            clType.setChildTypes(Arrays.asList(value.getValue0().getClType(), value.getValue1().getClType(),
+                    value.getValue2().getClType()));
+        } else {
+            clType.setChildTypes(new ArrayList<>());
+        }
     }
 
     @Override

--- a/src/test/java/com/casper/sdk/service/CasperServiceTests.java
+++ b/src/test/java/com/casper/sdk/service/CasperServiceTests.java
@@ -2,7 +2,6 @@ package com.casper.sdk.service;
 
 import com.casper.sdk.exception.CasperClientException;
 import com.casper.sdk.exception.DynamicInstanceException;
-import com.casper.sdk.identifier.block.BlockIdentifier;
 import com.casper.sdk.identifier.block.HashBlockIdentifier;
 import com.casper.sdk.identifier.block.HeightBlockIdentifier;
 import com.casper.sdk.identifier.global.BlockHashIdentifier;
@@ -43,7 +42,12 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 /**
  * Unit tests for {@link CasperService}
@@ -200,9 +204,8 @@ public class CasperServiceTests extends AbstractJsonRpcTests {
         assertTrue(deployData.getDeploy().getSession() instanceof StoredContractByHash);
         assertTrue(deployData.getExecutionResults().get(0).getResult() instanceof Success);
         assertTrue(deployData.getDeploy().getPayment() instanceof ModuleBytes);
-        assertTrue(deployData.getDeploy().getSession() instanceof StoredContractByHash);
-        String tmp = ((StoredContractByHash) deployData.getDeploy().getSession()).getHash();
-        assertEquals("ccb576d6ce6dec84a551e48f0d0b7af89ddba44c7390b690036257a04a3ae9ea", tmp);
+        String sessionHash = ((StoredContractByHash) deployData.getDeploy().getSession()).getHash();
+        assertEquals("ccb576d6ce6dec84a551e48f0d0b7af89ddba44c7390b690036257a04a3ae9ea", sessionHash);
     }
 
 
@@ -314,11 +317,19 @@ public class CasperServiceTests extends AbstractJsonRpcTests {
     void getCasperClientExceptionExceptionBlockNotKnown() {
         String expectedMessage = "Invalid params (code: -32602)";
 
-        BlockIdentifier blockIdentifier = new HashBlockIdentifier("abc");
-
         CasperClientException casperClientException = assertThrows(CasperClientException.class,
                 () -> casperServiceMainnet.getDeploy("abc"));
 
         assertEquals(expectedMessage, casperClientException.getMessage());
+    }
+
+    @Test
+    void getDeployMainnetWithCLValueListEmpty() {
+        assertDoesNotThrow(() -> casperServiceMainnet.getDeploy("2f430047c90597ef30467bb5393081e83a69c76ae215b1a89e01918748831a60"));
+    }
+
+    @Test
+    void getDeployMainnetWithCLValueMapEmpty() {
+        assertDoesNotThrow(() -> casperServiceMainnet.getDeploy("f8c3d8a9bc4b2a07928c9eeaa8bac41ec54b13cc767eb25509770b333f0f77e9"));
     }
 }

